pep-9999.rst

```diff
@@ -0,0 +1,294 @@
+             24 JUMP_BACKWARD            6 (to 14)
+        >>   26 END_FOR
+             28 SWAP                     2
+             30 STORE_FAST               1 (x)
+             32 RETURN_VALUE
+
+There is no longer a separate code object, nor creation of a single-use function
+object, nor any need to create and destroy a Python frame.
+
+Isolation of the ``x`` iteration variable is achieved by the combination of the
+new ``LOAD_FAST_AND_CLEAR`` opcode at offset ``6``, which saves any outer value
+of ``x`` on the stack before running the comprehension, and ``30 STORE_FAST``,
+which restores the outer value of ``x`` (if any) after running the
+comprehension.
+
+If the comprehension accesses variables from the outer scope, inlining avoids
+the need to place these variables in a cell, allowing the comprehension (and all
+other code in the outer function) to access them as normal fast locals instead.
+This provides further performance gains.
+
+This PEP only proposes to inline comprehensions occurring inside functions,
+where fast-locals (``LOAD_FAST/STORE_FAST``) are used. Module-level
+comprehensions will continue to create and call a function. Module-level
+comprehensions are generally called only once (when the module is imported), so
+optimizing the performance of calling them is unnecessary, and inlining them
+would require separate code paths in the compiler to handle a module global
+namespace dictionary instead of fast-locals.
+
+
+Backwards Compatibility
+=======================
+
+Comprehension inlining will cause the following visible behavior changes. No
+changes in the standard library or test suite were necessary to adapt to these
+changes in the implementation, suggesting the impact in user code is likely to
+be minimal.
+
+Specialized tools depending on undocumented details of compiler bytecode output
+may of course be affected in ways beyond the below, but these tools already must
+adapt to bytecode changes in each Python version.
+
+locals() includes outer variables
+---------------------------------
+
+Calling ``locals()`` within a comprehension will include all locals of the
+function containing the comprehension. E.g. given the following function::
+
+  def f(lst):
+      return [locals() for x in lst]
+
+Calling ``f([1])`` in current Python will return::
+
+  [{'.0': <list_iterator object at 0x7f8d37170460>, 'x': 1}]
+
+Where ``.0`` is the internal implementation detail of the synthetic sole
+argument to the comprehension "function."
+
+Under this PEP, it will instead return::
+
+  [{'lst': [1], 'x': 1}]
+
+Now including the outer ``lst`` variable as a local, and eliminating the
+synthetic ``.0``.
+
+no comprehension frame in tracebacks
+------------------------------------
+
+Under this PEP, a comprehension will no longer have its own dedicated frame in a
+stack trace. For example, given this function::
+
+  def f():
+    def g():
+      raise RuntimeError("boom")
+
+    return [g() for x in [1]]
```

Comment:
I found the structure here with the nested indent to be a little less clear and more distracting than two separate functions, when it doesn't appear to be needed to get the point across. Could it be simplified to

```suggestion
  def g():
      raise RuntimeError("boom")

  def f():
      return [g() for x in [1]]
```

instead (and normalized to standard 4-space indents used throughout the rest of the PEP)?